plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -0,0 +1,1072 @@
+        buf->data_size += strlen(stream->sequence_token);
+    }
+}
+
+/* sorts events, constructs a put payload, and then sends */
+int send_log_events(struct flb_cloudwatch *ctx, struct cw_flush *buf,
+                    struct log_stream *stream) {
+    int ret;
+    int offset;
+    int i;
+    struct event *event;
+
+    if (buf->event_index == 0) {
+        return 0;
+    }
+
+    /* events must be sorted by timestamp in a put payload */
+    qsort(buf->events, buf->event_index, sizeof(struct event), compare_events);
+
+retry:
+    stream->newest_event = 0;
+    stream->oldest_event = 0;
+
+    offset = 0;
+    ret = init_put_payload(ctx, buf, stream, &offset);
+    if (ret < 0) {
+        flb_plg_error(ctx->ins, "Failed to initialize PutLogEvents payload");
+        return -1;
+    }
+
+    for (i = 0; i < buf->event_index; i++) {
+        event = &buf->events[i];
+        ret = write_event(ctx, buf, event, &offset);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "Failed to write log event %d to "
+                          "payload buffer", i);
+            return -1;
+        }
+        if (i != (buf->event_index - 1)) {
+            if (!try_to_write(buf->out_buf, &offset, buf->out_buf_size,
+                              ",", 1)) {
+                flb_plg_error(ctx->ins, "Could not terminate log event with ','");
+                return -1;
+            }
+        }
+    }
+
+    ret = end_put_payload(ctx, buf, &offset);
+    if (ret < 0) {
+        flb_plg_error(ctx->ins, "Could not complete PutLogEvents payload");
+        return -1;
+    }
+
+    flb_plg_debug(ctx->ins, "Sending %d events", i);
+    ret = put_log_events(ctx, buf, stream, (size_t) offset);
+    if (ret < 0) {
+        flb_plg_error(ctx->ins, "Failed to send log events");
+        return -1;
+    } else if (ret > 0) {
+        goto retry;
+    }
+
+    return 0;
+}
+
+/*
+ * Processes the msgpack object, sends the current batch if needed
+ */
+int add_event(struct flb_cloudwatch *ctx, struct cw_flush *buf,
+              struct log_stream *stream,
+              const msgpack_object *obj, struct flb_time *tms)
+{
+    int ret;
+    struct event *event;
+    int new_len;
+    size_t size;
+    int retry_add = FLB_FALSE;
+    int event_bytes = 0;
+
+    if (buf->event_index == 0) {
+        /* init */
+        reset_flush_buf(ctx, buf, stream);
+    }
+
+    /* re-alloc event buffer if needed */
+    if ((buf->event_index + 1) >= buf->events_capacity) {
+        new_len = MAX_EVENTS_PER_PUT;
+        size = sizeof(struct event) * new_len;
+        flb_plg_debug(ctx->ins, "Increasing event buffer to %d", new_len);
+        buf->events = flb_realloc(buf->events, size);
+        if (!buf->events) {
+            flb_errno();
+            return -1;
+        }
+        buf->events_capacity = new_len;
+    }
+
+retry_add_event:
+    ret = process_event(ctx, buf, obj, tms);
+    if (ret < 0) {
+        return -1;
+    } else if (ret > 0) {
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions